### PR TITLE
sort-pushdown v1 poc

### PR DESCRIPTION
**Pull request by yamatattsu**
Hi! Laurenz,

I'm interested in oracle_fdw, and I'd like to contribute to develop it.
I created a patch of sort-pushdown v1.  It is POC.
The patch is used following commit as a reference. 
Therefore, It has some  discussing points, such as design, implementation.

     V1: postgres_fdw: Add ORDER BY to some remote SQL queries.
           https://github.com/postgres/postgres/commit/f18c944b6137329ac4a6b2dce5745c5dc21a8578

Could you watch my code?

Thanks,
Tatsuro Yamada

### CODE REVIEW

**Comment by laurenz** [comment]
I've finally got to review your pull request.

In oracle_fdw, all the work for constructing a remote SELECT is done in `oracleGetForeignRelSize`, so I think it would be best to keep things that way.
Here are my suggestions for how to refactor your patch so that it fits better and avoids unnecessary code duplication:

1. `find_em_expr_for_rel` is obviously ok the way it is, we can just keep it.

2. `struct OracleFdwState` should get a new field `orderedQuery` (or similar) that can contain the query with an appended ORDER BY clause and a field `usable_pathkeys`.

3. [The code](https://github.com/laurenz/oracle_fdw/pull/80/commits/bf9005e7a884895680072bcd64bc122fdb0fbc37#diff-72b991171844f1b822159147a3579f62R889) that determines whether ORDER BY can be pushed down should be moved from `oracleGetForeignPaths` to `oracleGetForeignRelSize`.

4. `is_foreign_expr` and `deparseExpr` are not necessary and should be replaced with a call to `getOracleWhereClause`, roughly like this:

		StringInfoData order_by;
		initStringInfo(&order_by);
		
		foreach(lc, root->query_pathkeys)
		{
			char *sort_clause;
			bool first_sortkey = true;
		
			/*
			 * is_foreign_expr would detect volatile expressions as well, but
			 * ec_has_volatile saves some cycles.
			 */
			if (!pathkey_ec->ec_has_volatile &&
				(em_expr = find_em_expr_for_rel(pathkey_ec, baserel)) &&
				(sort_clause = getOracleWhereClause(fdwState->session, baserel, em_expr, ...)))
			{
				fdwState->usable_pathkeys = lappend(usable_pathkeys, pathkey);
				
				if (first_sortkey)
					first_sortkey = false;
				else
					appendStringInfo(&order_by, ", ");
				appendStringInfo(&order_by, "%s", sort_clause);
				/* append ASC, DESC and NULLS FIRST/LAST */
			}
			else
				/* clear order_by and exit loop */
		}

5. Change `createQuery` so that it does not prepend `SELECT /* md5 */` to the query string and does not add `FOR UPDATE`.

6. In `oracleGetForeignRelSize`, where `fdwState->query` is constructed, change the code so that the following is done:

    6.1. The result of `createQuery()` is assigned to a string variable, e.g. `query`.
    6.2. `fdwState->query` is constructed by prepending `SELECT /* md5 */` to `query` and optionally appending `FOR UPDATE`.
    6.3. If `order_by` is initialized, set `fdwState->orderedQuery` to `SELECT /* md5 */` append `query` append ` ORDER BY ` append `order_by` append (optionally) `FOR UPDATE`.

7. `getOracleWhereClause` will have to be modified to serve this additional purpose, probably by adding another parameter `bool for_order_by`.
For example, a `T_Var` of type `BOOLOID` should be replaced by the Oracle column name and not by `(colname <> 0)`.

8. In `oracleGetForeignPaths`, construct a second path if appropriate using `fdwState->orderedQuery` and `fdwState->usable_pathkeys`.

Does that make sense?
Please don't hesitate to ask if any of the above is unclear.

**Comment by laurenz** [comment]
Using `getOracleWhereClause` instead of `is_foreign_expr` should also have the advantage that more complicated ORER BY expressions could be pushed down, or is there an error in my reasoning?

There would have to be some special case processing to keep string expressions from appearing in ORDER BY clauses.

**Comment by laurenz** [comment]
I slept over it, and have got some additional ideas how to simplify things.

You didn't do anything about costs for the new plan yet, so the cost is the same as for the unsorted plan. That will make PostgreSQL always pick the sorted plan, since it thinks it can get away cheaper that way.
If that is an oversight, it is a good one, since there is no good way to reliably determine costs for the sort in Oracle anyway. We could just use a factor of 1.2 as postgres_fdw does in this case, but I don't think that will be very useful.
We could just say that (as it is now) the sorting should *always* take place on the Oracle side. I believe that sorting in Oracle does not perform worse than sorting in PostgreSQL, and it might perform much better if an index can be used.

So I propose that we just generate the sorted plan if we can.

That way we don't need `fdwState->orderedQuery`, and instead of factoring the `SELECT /* m5 */` out from `createQuery` we could leave it in and also have the ORDER BY clause be generated inside that function.
(I consider renaming `getOracleWhereClause` to something like `deparseExpression` since that seems more appropriate now.)

What do you think of that?

**Comment by yamatattsu** [comment]
First off, I answer your last comment.

I didn't implement remote cost estimation for determining whether do sort-pushdown or do not sort-pushdown. Because I have no idea of it. 
Therefore, I agree with your suggestion which the sorting should always take place on the Oracle side.
If we get other suggestion such as a complaint, We can use GUC parameter to switch and will solve it.
Then I also agree to rename the getOracleWhereClause to more appropriate.

**Comment by yamatattsu** [comment]
Next, I answer your comments(from 1. to 8. ).

1. Keep on find_em_expr_for_rel.
  -> OK!

2. Get a new field OracleFdwState->orderedQuery.
  -> By your the last comment, I don't create it. Is it OK?

3. A processing which determining whether pushdown move from oracleGetForeignPaths to oracleGetForeignRelSize.
  -> In the case of according to the policy of your last comment,
     if we can get the path_keys from baserel or fdwstate, the processing can move to oracleGetForeignRelSize, I think.
But I'm not sure can we get the path_keys from baserel or fdwstate on now. 

4. is_foreign_expr and deparseExpr should be replaced getOracleWhereClause.
  -> I'll try it, but it will take some time to it.
 
5. Change createQuery.
  -> According to your last comment, I don't change it.

6. Change oracleGetForeignRelSize.
  -> I'd like to start when I finished an investigation of 3.

7. Modify getOracleWhereClause.
  -> I'm not sure your comment. My question is this.
    * What's happen when we didn't change T_Var?

8. Create 2nd path in oracleGetForeignPaths
  -> I don't know well... I suppose We need a path to get only sorting plan. so Do we need the 2nd path? Is my understanding right?

**Comment by laurenz** [comment]
- Yes, if we generate only one path in `oracleGetForeignRelSize`, we don't need `OracleFdwState->orderedQuery` any more.

- Looking at your code, `pathkeys` is found using the parameters `root` and `baserel` of `oracleGetForeignPaths`.
  `oracleGetForeignRelSize` has these parameters as well, so it should be possible to move the code there to keep query generation in one place.

- My explanation for point 7 has been too short.
  I mean that we can pass the `ORDER BY` expressions to `getOracleWhereClause` (yes, it should be renamed then) to deparse them for Oracle.
  The comment about boolean column references means the following:
    - When an Oracle `NUMBER` column is mapped to PostgreSQL `boolean`, `getOracleWhereClause` will replace a column reference with `("BOOLCOL" <> 0)` so it can be used in a logical expression.
    - If we repurpose `getOracleWhereClause` to deparse `ORDER BY` expressions, this could cause a syntax error. So we will have to process differently for `WHERE` and for `ORDER BY` expressions.
    - A possible solution would be to emit different things for `WHERE` clauses and for `ORDER BY` expressions. We might also just give up (return `NULL`) in that case, because I have never seen `ORDER BY coolcol` in a query anyway.

- Ad 8.: we don't need the second path any more, as we agreed.

Let me know if I have answered all your questions.

**Comment by yamatattsu** [comment]
The patch has not completed, but I'd like to show the current progress status of revising the patch.

2.
I create a new parameter: OracleFdwState->usable_pathkeys.
It is able to hand over usable_pathkeys from oracleGetForeignRelSize() to oracleGetForeignPaths().

3.
>Looking at your code, pathkeys is found using the parameters root and >baserel of oracleGetForeignPaths.
>oracleGetForeignRelSize has these parameters as well, so it should be >possible to move the code there to keep query generation in one place.

I had forgotten it.
I moved the code that determines whether ORDER BY can be pushded down from oracleGetForeignPaths() to racleGetForeignRelSize().

In oracleGetForeignRelSize(),
If there is a root->query_pathkeys and It's OK to pushdown, then deparse.   

4.
I removed is_foreign_expr and deparseExpr from the code, and replaced with a call to getOracleWhereClause. I used your code sample partly.
I added ORDER BY clause to query text in createQuery().

7.
>I mean that we can pass the ORDER BY expressions to getOracleWhereClause (yes, it should be renamed then) to deparse them for Oracle.
>The comment about boolean column references means the following:

Thanks for an explanation.
I'm still thinking How to change the code.
Then I added a regression test in oracle_fdw.sql.
We have to select a data type which can pushdown based on a difference of collation.
Actually, I tried pushdown all data type, but some of data types (text, bytea and boolean) got an error. And It's not fixed. below is error. These sql are not included in oracle_fdw.sql.

* EXPLAIN (VERBOSE on, COSTS off) SELECT * FROM typetest1 ORDER BY lc ;
psql:sort_pd_error.sql:2: ERROR:  error describing query: OCIStmtExecute failed to describe remote query
DETAIL:  ORA-00932: inconsistent datatypes: expected - got CLOB

* EXPLAIN (VERBOSE on, COSTS off) SELECT * FROM typetest1 ORDER BY b  ;
psql:sort_pd_error.sql:4: ERROR:  error describing query: OCIStmtExecute failed to describe remote query
DETAIL:  ORA-00907: missing right parenthesis

* EXPLAIN (VERBOSE on, COSTS off) SELECT * FROM typetest1 ORDER BY lr ;
psql:sort_pd_error.sql:5: connection to server was lost
LOG:  server process (PID 8826) was terminated by signal 11: Segmentation fault
DETAIL:  Failed process was running: EXPLAIN (VERBOSE on, COSTS off) SELECT * FROM typetest1 ORDER BY lr ;

**Comment by laurenz** [comment]
Thank you for your work!

Some comments:
- Please avoid white space change and re-indentation like in `oracleGetForeignPaths` if possible &ndash; it will make the patch much easier to read.
- `OracleFdwState.usable_pathkeys` should be initialized to `NULL` in `getFdwState`.  
  Then you don't need the test `if (fdwState->usable_pathkeys == NULL)` in `oracleGetForeignPaths` at all, which will make the code simpler.
- It is a bit wasteful and inelegant to call `getOracleWhereClause` twice, once during the check if `ORDER BY` expressions can be pushed down in `oracleGetForeignRelSize`, and the second time in `appendOrderByClause` when the actual query string is constructed.
  I think it would be better to move the check into `createQuery`.
  Looking at the function's signature, I think it might be simpler just to pass the whole `OracleFdwState *` instead of so many of its fields, then it would be easy to populate `usable_pathkeys` in `createQuery`.
- As you noted, there are some data types that cause problems:
  - Oracle `CLOB`: Even using sqlplus I get the error:  
    `ORA-00932: inconsistent datatypes: expected - got CLOB`  
    So I guess we shouldn't try to push down sorts on `CLOB`s.
  - Oracle `RAW`: regardless if matched to `bytea` or `uuid`, trying to `ORDER BY` such a column causes a crash. I did not research why, this will need some investigation. It works fine with sqlplus!
  - Oracle `BLOB`: I get a crash here too. With sqplus I get (similar to `CLOB`):  
    `ORA-00932: inconsistent datatypes: expected - got BLOB`  
    That shoudln't even be attempted.
  - PostgreSQL `boolean`: As expected, the query ends up like  
    `... ORDER BY ("B" <> 0) ASC NULLS LAST`  
    which causes a syntax error.  
    We can either push down `ORDER BY "B"` or just skip boolean columns &ndash; who really cares to order by a boolean?

  `ORDER BY` boolean could be detected by examining the type of the expression, but the others will be a little more complicated.
  We should definitely try to get `RAW` to work, there must be a silly bug somewhere.
  About `BLOB` and `CLOB`: we could modify `getOracleWhereClause` so that it returns `NULL` if it encounters a variable of type `CLOB` or `BLOB`, a good place might be around [here](https://github.com/yamatattsu/oracle_fdw/blob/sort_pushdown_v1_poc/oracle_fdw.c#L2892). That would require an argument to `getOracleWhereClause` so that it knows whether it works on a `WHERE` or an `ORDER BY` expression.

**Comment by yamatattsu** [comment]
Thanks for youre thoughtful comments.
I revised the code based on your comments.

- white space and indentation

 -> I see, I'll be more careful in future.

- OracleFdwState.usable_pathkeys should be initialized to NULL in getFdwState.

 -> I fixed it.

- It is a bit wasteful and inelegant to call getOracleWhereClause twice...

 -> I fixed it.
   I created the actual query string in createQuery() for order by clause. And I changed the first parameter "struct oracleFdwState *fdwState" in getOracleWhereClause().
I also added "isSort" parameter because I'd like to control based on caller.

- there are some data types that cause problems...

 -> I experimentally added below code in  getOracleWhereClause().
It seems like good.

`if(!isSort){`
`    ...`
`}`
`else`
`{`
`       if(variable->vartype == BOOLOID`
`                                || oratype == ORA_TYPE_CLOB)`
`                            return NULL;`
`}`

File: oracle_fdw.c [hunk 1]
```diff
@@ -705,7 +703,7 @@ oraclePlanForeignScan(Oid foreigntableid,
 	fdwState = getFdwState(foreigntableid, &plan_costs, NULL);
 
 	/* construct Oracle query and get the list of parameters and actions for RestrictInfos */
-	fdwState->query = createQuery(fdwState->session, baserel, false, fdwState->oraTable, &(fdwState->params), &(fdwState->pushdown_clauses), usable_pathkeys);
+	fdwState->query = createQuery(fdwState->session, baserel, false, fdwState->oraTable, &(fdwState->params), &(fdwState->pushdown_clauses), NIL);
```
Comment (laurenz) [inline oracle_fdw.c]: You forgot to change the function signature here.
I guess you didn't build with PostgreSQL 9.1...

**Comment by laurenz** [comment]
I have created a new branch 'sort_pushdown' and merged your changes.

Apart from the one omission I commented in your patch (function signature), I think that the handling in `getOracleWhereClause` is far from being in a working shape.

I'll whack your patch around a little to suit my ideas about code formating better, also I'll try to come up with something for `getOracleWhereClause`.

Finally I'd like to rename `getOracleWhereClause` to something more fitting.

Please stand by...

**Comment by laurenz** [comment]
I just pushed 3008ce05332ab9db6cbb96e5d982a1defdb0fc4a which should handle the decision what to push down and what not.

I realized that the decision should depend only on the type of the whole expression, so the `isSort` parameter proved unnecessary after all. Sorry for causing you the trouble of adding it everywhere&hellip;

If that looks ok with you, I'll work on the regression tests next.

**Comment by yamatattsu** [comment]
Thanks for improving the patch.
I respond to your questions and comments with the following.

>I guess you didn't build with PostgreSQL 9.1...

Oops, sorry. I have only focused on 9.6. So, I 've forgotten it.

>I have created a new branch 'sort_pushdown' and merged your changes.
>...
>Please stand by...

Okey, I will get that ready.

>I just pushed 3008ce0 which should handle the decision what to push down and what not.
>
>I realized that the decision should depend only on the type of the whole expression, so the isSort >parameter proved unnecessary after all. Sorry for causing you the trouble of adding it everywhere…
>
>If that looks ok with you, I'll work on the regression tests next.

Great. I watched your changes and I agree with you because It looks OK to me.
Then, It's no problem about removing the isSort parameter.

**Comment by laurenz** [comment]
I have merged the code, ORDER BY pushdown will be in the next release.

Thank you for your contribution!
